lib/vector-common/src/request_metadata.rs

```diff
@@ -1,16 +1,180 @@
+        Self::Untagged { size: value }
+    }
+}
+
+impl<'a> Add<&'a RequestCountByteSize> for RequestCountByteSize {
+    type Output = RequestCountByteSize;
+
+    fn add(self, other: &'a Self::Output) -> Self::Output {
+        match (self, other) {
+            (
+                RequestCountByteSize::Tagged { sizes: mut us },
+                RequestCountByteSize::Tagged { sizes: them },
+            ) => {
+                for (key, value) in them {
+                    match us.get_mut(key) {
+                        Some(size) => *size += *value,
+                        None => {
+                            us.insert(key.clone(), *value);
+                        }
+                    }
+                }
+
+                Self::Tagged { sizes: us }
+            }
+
+            (
+                RequestCountByteSize::Untagged { size: us },
+                RequestCountByteSize::Untagged { size: them },
+            ) => RequestCountByteSize::Untagged { size: us + *them },
+
+            // The following two scenarios shouldn't really occur in practice, but are provided for completeness.
+            (
+                RequestCountByteSize::Tagged { mut sizes },
+                RequestCountByteSize::Untagged { size },
+            ) => {
+                match sizes.get_mut(&(OptionalTag::Specified(None), OptionalTag::Specified(None))) {
+                    Some(empty_size) => *empty_size += *size,
+                    None => {
+                        sizes.insert(
+                            (OptionalTag::Specified(None), OptionalTag::Specified(None)),
+                            *size,
+                        );
+                    }
+                }
+
+                Self::Tagged { sizes }
+            }
+            (RequestCountByteSize::Untagged { size }, RequestCountByteSize::Tagged { sizes }) => {
+                let mut sizes = sizes.clone();
+                match sizes.get_mut(&(OptionalTag::Specified(None), OptionalTag::Specified(None))) {
+                    Some(empty_size) => *empty_size += size,
+                    None => {
+                        sizes.insert(
+                            (OptionalTag::Specified(None), OptionalTag::Specified(None)),
+                            size,
+                        );
+                    }
+                }
+
+                Self::Tagged { sizes }
+            }
+        }
+    }
+}
+
 /// Metadata for batch requests.
-#[derive(Clone, Copy, Debug, Default)]
+#[derive(Clone, Debug, Default)]
 pub struct RequestMetadata {
     /// Number of events represented by this batch request.
     event_count: usize,
     /// Size, in bytes, of the in-memory representation of all events in this batch request.
     events_byte_size: usize,
     /// Size, in bytes, of the estimated JSON-encoded representation of all events in this batch request.
-    events_estimated_json_encoded_byte_size: JsonSize,
+    events_estimated_json_encoded_byte_size: RequestCountByteSize,
```

Comment:
It feels a little weird that we have this field which might be grouped, might not, while the others are never grouped.

Not really sure, as I type this comment, what a better solution would be but it does feel like it adds on more complexity to what was already (admittedly, given that I designed it) an already somewhat opaque type.

Reply:
Yes, it is unfortunate primarily because this struct can no longer be `Copy`, 
- which means we now have to have a `take_metadata` function on the  `MetaDescriptive` trait so that we don't have to `clone` the object,
- which then means that we have to be very careful we don't access the metadata twice on that object because the second time the real values will no longer be there.

A very elusive bug could creep in.

I'm all ears for a better solution.